doc/sonic-application-extension/tpcm_app_ext.md

```diff
@@ -0,0 +1,467 @@
+
+##### CPU Resource
+- By default, for each TPC docker, the CPU resource limit is restricted to 20%. 
+
+##### Memory Limits
+- By default, for each TPC docker, the memory limit is restricted to 512M
+
+
+
+### SAI API 
+
+Not applicable
+
+### Configuration and management 
+
+<!-- omit in toc -->
+#### CLI Enhancements
+
+The SONiC Package Manager is another executable utility available in base SONiC OS called *sonic-package-manager* or abbreviated to *spm*. This would be extended to support these new TPCM capabilities. The command line interfaces are given bellow:
+
+<!-- omit in toc -->
+#### CLI
+
+<!-- omit in toc -->
+#### TPC Installation
+
+This section shows the additional options that would be added for TPC installation in the sonic-package-manager CLI
+
+```
+admin@sonic:~$ sudo sonic-package-manager install --help
+
+Usage: sonic-package-manager install [OPTIONS] [PACKAGE_EXPR]
+
+  Install SONiC package.
+
+Options:
+  --from-tarball-scp                Install using the tarball from SCP source
+  --from-tarball-sftp               Install using the tarball from SFTP source
+  --from-tarball-url                Install using the tarball from HTTPS source
+  --use-local-manifest              Use locally created manifest file
+  --name                            custom name for tpcm package
+  ```
+
+<!-- omit in toc -->
+###### Examples
+
+```
+admin@sonic:~$ sudo sonic-package-manager install httpd:latest --name my_httpd
+```
+
+Install from url:
+```
+admin@sonic:~$ sudo sonic-package-manager install --from-tarball-url https://tpc.local-server.com/home/tpcs/httpd.tar.gz --name=my_url_httpd
+```
+
+Install from scp:
+```
+admin@sonic:~$ sudo sonic-package-manager install --from-tarball-scp /home/tpcs/httpd.tar.gz --host 10.171.112.156 --username user1 --password pass123 --name my_scp_httpd
+```
+
+Install from sftp:
+```
+admin@sonic:~$ sudo sonic-package-manager install --from-tarball-sftp /home/tpcs/httpd.tar.gz --host 10.171.112.156 --username user1 --password pass123 --name my_sftp_httpd
+```
+Install from local FS:
+```
+admin@sonic:~$ sudo sonic-package-manager install --from-tarball /usb1/tpcs/httpd.tar.gz --name my_local_httpd
+```
+#### TPC Uninstallation
+
+sonic-package-manager uninstall option shall be used to uninstall the TPC
+
+
+<!-- omit in toc -->
+###### Examples
+
+```
+admin@sonic:~$ sudo sonic-package-manager uninstall my_scp_httpd
+```
+
+
+#### TPC Resource limits
+
+##### CPU Resource limit
+
+As a policy, we will allocate only 20% of the CPU for TPCs.After installation, the default CPU values would be:
+
+--cpu-period=100000 --cpu-quota=20000
+
+These parameters are restricted to be passed as args to the TPC install.
+
+##### Memory Resource limit
+
+Default Memory limit for a TPC would be 512M
+
+
+<!-- omit in toc -->
+###### Examples
+
+```
+admin@sonic:~$ sudo sonic-package-manager manifests create my_node_exporter --tpcm --container memory 50m  
```

Comment:
How do we make sure that TPC is not hijacking any of the control package meant for processing by SONiC system? do we have any validation in place?

Reply:
TPC docker wont run in "privileged" mode and runs with the default security settings and won't have escalated privileges.

Reply:
If we can skip '--tpcm' option in the command, would be good.

Reply:
Skipped as suggested